core/server/master/src/main/java/alluxio/master/file/activesync/ActiveSyncer.java

```diff
@@ -74,29 +86,47 @@ public void heartbeat() {
     }
 
     try {
-      UfsManager.UfsClient ufsclient = mMountTable.getUfsClient(mMountId);
+      UfsManager.UfsClient ufsclient = Objects.requireNonNull(mMountTable.getUfsClient(mMountId));
       try (CloseableResource<UnderFileSystem> ufsResource = ufsclient.acquireUfsResource()) {
         UnderFileSystem ufs = ufsResource.get();
         if (ufs.supportsActiveSync()) {
           SyncInfo syncInfo = ufs.getActiveSyncInfo();
           // This returns a list of ufsUris that we need to sync.
           Set<AlluxioURI> ufsSyncPoints = syncInfo.getSyncPoints();
           // Parallelize across sync points
-          List<Callable<Void>> tasksPerSyncPoint = new ArrayList<>(ufsSyncPoints.size());
+          List<CompletableFuture<Long>> tasksPerSync = new ArrayList<>();
           for (AlluxioURI ufsUri : ufsSyncPoints) {
-            tasksPerSyncPoint.add(() -> {
+            tasksPerSync.add(CompletableFuture.supplyAsync(() -> {
               processSyncPoint(ufsUri, syncInfo);
-              return null;
-            });
+              return syncInfo.getTxId();
+            }, mSyncManager.getExecutor()));
           }
-          mSyncManager.getExecutor().invokeAll(tasksPerSyncPoint);
           // Journal the latest processed txId
-          mFileSystemMaster.recordActiveSyncTxid(syncInfo.getTxId(), mMountId);
+          CompletableFuture<Void> syncTask =
+              CompletableFuture.allOf(tasksPerSync.toArray(new CompletableFuture<?>[0]))
+              .thenRunAsync(() -> mFileSystemMaster
+                      .recordActiveSyncTxid(syncInfo.getTxId(), mMountId),
+                  mSyncManager.getExecutor());
+          while (!mSyncTasks.offer(syncTask)) {
```

Comment:
If the first task in the queue is slow for some reason, would that hold up all the rest of the processing, even if the rest of the tasks have completed?

Reply:
its possible that it holds up removing tasks from the queue, though for this to happen is would require filling the incremental sync queue which right now holds up to 32 items. Given the standard heartbeat interval of 30 seconds that means this would need to take 16 minutes to finish the incremental sync.

I can try improving this by using a set to iterate over all of the elements and removing any that are finished, but would require iterating over all futures. I think for the size of the futures added to this queue that is a reasonable solution

Reply:
updated